web/app/scripts/views/map/layers-controller.js

```diff
@@ -116,25 +121,47 @@
                     $rootScope.$broadcast('driver.views.map:filterdrawn', null);
                 });
 
+                ctl.map.on('zoomend', function() {
+                    MapState.setZoom(ctl.map.getZoom());
+                });
+
+                ctl.map.on('moveend', function() {
+                    MapState.setLocation(ctl.map.getCenter());
+                });
+
                 // TODO: Find a better way to ensure this doesn't happen until filterbar ready
                 // (without timeout, filterbar components aren't ready to listen yet)
                 // add filtered overlays
                 // this will trigger `driver.filterbar:changed` when complete
                 $timeout(FilterState.restoreFilters, 1000);
+            }).then(function() {
+                if (MapState.getLocation() && MapState.getZoom()) {
+                    ctl.map.setView(MapState.getLocation(), MapState.getZoom());
+                }
+
+                if (MapState.getFilterGeoJSON()) {
+                    var layer =  L.geoJson(MapState.getFilterGeoJSON());
+                    layer.setStyle(filterStyle);
+                    ctl.editLayers.addLayer(layer);
+                }
             });
         };
 
-        function filterShapeCreated(event) {
+        function filterShapeCreated(layer) {
             // TODO: is the shape type useful info?
             //var type = event.layerType;
-
-            var layer = event.layer;
-
             ctl.editLayers.clearLayers();
-            $rootScope.$broadcast('driver.views.map:filterdrawn', null);
+
+            layer.setStyle(filterStyle);
             ctl.editLayers.addLayer(layer);
+            $rootScope.$broadcast('driver.views.map:filterdrawn');
+
+            // Use GeoJSON instead of a normal layer - theres a strange bug likely stemming from
+            //  race conditions on the Leaflet Map object otherwise
```

Comment:
Yes, but leaflet-draw layers haven't been saved and then applied to a newly initialized map until this PR. Here's what I was seeing from a high level:
1. I save a layer (e.layer from leaflet-draw's event)
2. I attempt to map.addLayer it
3. A stack trace bursts out from deep within the bowels of Leaflet complaining of undefined properties

I'm just creating a layer afresh - that's literally the only difference - so I have a very hard time believing that this somehow involves code we wrote. Now, perhaps I just fundamentally misunderstand the leaflet ontology - if that is so, I'd definitely like to be corrected - but with the concepts I've got at my disposal, I just don't know how else to think about this problem. 

Would you like to take a look before I merge this in? Like I said, if there's something I'm just ignorant of, I'd like to be corrected